### DISCUSSION TITLE
Enhancements + Cleaned up code

### DISCUSSION THREAD

**Discussion by Green0-0** [VsonicV/es-fine-tuning-paper#22] [Show and tell]
Hi! I’ve replicated the experiments from the paper and cleaned up the codebase.
My setup use a mini-batch of 300 on the full dataset, with 10% held out for validation (evaluated every 10 steps). I also adjusted the prompt formatting to use the official prompt template, and the format reward function has been made slightly more accurate. I used the non-iid version of the perturbations, but admittedly iid would've been a better choice. All runs are tested with 4xa5000s for 250 steps, taking around a day for each run.

<img width="1154" height="315" alt="image" src="https://github.com/user-attachments/assets/7fa98a60-5e04-44a0-b2b3-85f99df1de89" />
<img width="1159" height="331" alt="image" src="https://github.com/user-attachments/assets/c08ab3c7-4c4b-42b2-988c-df33f2d83b6b" />

The results line up with the results of the paper. Interestingly, even after halving the population, the method still performs competitively. Increasing the population from 12 to 28 roughly doubled the runtime from 2 minutes a step to 4 minutes a step.
The validation scores are almost 1:1 with the training scores. In fact, with this particular validation split, the validation scores are better than the training scores. 

With the population and batch sizes fixed, I tried a bunch of common enhancements, all with a learning rate of 0.0015 (3x):
<img width="1165" height="315" alt="image" src="https://github.com/user-attachments/assets/d05b3d4a-9590-49c5-bcb8-12d7e40df854" />
<img width="1154" height="321" alt="image" src="https://github.com/user-attachments/assets/6bc3fa53-28a9-4db3-919f-8f514c69f023" />
Mirroring had no impact on the quality of the model (actually, it diverged at the end but that might just be the high LR). Tripling the learning rate improved convergence speed, but didn't improve the final outcome. A cosine LR scheduler also changed the convergence curve without changing the final performance. Perhaps I did not train the model for long enough, or countdown is simply too easy.

Lastly, I experimented with momentum. By caching the seeds and their associated weights, you can implement momentum with no overhead, which seemed very promising:
<img width="1159" height="311" alt="image" src="https://github.com/user-attachments/assets/b35ac0ed-70fe-42f6-aed9-718ee4fb8295" />
<img width="1160" height="329" alt="image" src="https://github.com/user-attachments/assets/1c48823c-bcb3-4cb5-ace0-3afa604a95b4" />
Unfortunately, momentum didn't behave consistently. In two runs, it sharply diverged after a few steps. Lowering the momentum coefficient to 0.5 seemed to fix this problem, but the performance did not increase at all. My suspicion is that the momentum buffer is storing too much noise from the previous steps, which accumulates and eventually overpowers the signal from the latest gradient. When this happens, the model diverges. 
I did not experiment with the momentum cutoff (where gradients past a certain step are simply dropped from the momentum buffer) which might also alleviate this problem. Someone with more compute than me should run a hyperparameter sweep here.

Considering that this model seems to be stuck at 60% performance on this dataset, I will move on and try a different model with a different dataset (perhaps qwen 3 base with https://huggingface.co/datasets/internlm/OREAL-RL-Prompts as the dataset).

All my code is at https://github.com/Green0-0/propagate along with a link to the graphs on wandb. I've also logged a lot more statistics such as runtime, standard deviation, response length, etc in case it'll serve a use.

**Comment by Green0-0** [comment]
Also, I think that an optimized LoRA implementation would speedup training significantly. I believe I was getting <1000t/s on qwen 3b when I tested on an A6000, but in theory you should be able to get over 10000 t/s (https://www.reddit.com/r/LocalLLaMA/comments/1mt2iev/gptoss20b_at_10000_tokenssecond_on_a_4090_sure/) with A6000s, H100s, etc. Additionally, LoRA might actually improve the performance of the algorithm by shrinking down the parameter space.
Regardless, I'll be following https://github.com/Bhoy1's work, and hopefully get it working myself in the coming weeks.

**Reply by Bhoy1** [reply]
Let me know if you every try another dataset and what you see, not opposed to trying one after I do this LoRA run on a batch of countdown

**Reply by Green0-0** [reply]
I have a run with qwen3 base and https://huggingface.co/datasets/internlm/OREAL-RL-Prompts (I kept only the integer answers) training.

The reason I chose the qwen3 base model is because the qwen3 instruct/thinking models already achieve ridiculous scores on datasets like AIME, so training on them would be pointless. It'll also be nice to see if the base model trains well, and what kind of behaviors it exhibits.

The updated code has been pushed already, for reference.

**Reply by Bhoy1** [reply]
Didn't the paper only run it with base models? " On smaller models such as Qwen2.5-0.5B and Llama-3.2-1B, RL showed almost no improvement, whereas ES steadily increased accuracy. On mid-sized models like Qwen2.5-1.5B, Qwen2.5-3B and Llama-3.2-3B, ES achieved substantial gains, reaching accuracy levels that RL never approached even in extended training. On larger models such as Qwen2.5-7B and Llama-3.1-8B,"

**Reply by Green0-0** [reply]
<img width="660" height="219" alt="image" src="https://github.com/user-attachments/assets/624c3d40-f1c0-46ab-8aa4-882cf30d2745" />

here is the chart from the paper, which cites qwen instruct

In the code as well here they use qwen instruct, I believe it is normal to use the instruct version as deepseek found that doing RL on the base model leads to unpredictable behaviors such as speaking in a different language.

**Reply by Bhoy1** [reply]
Okay on the conciseness task I posted, I just used the Qwen2.5-7B. I'll redo it with the Instruct version. Thanks for that catch, I was looking at that one Appendix figure when I was setting up the code and didn't check it against what they said earlier.

**Reply by VsonicV** [reply]
@Green0-0 is right. We were using the "instruct" version of these models. @Bhoy1

**Comment by VsonicV** [comment]
@Green0-0 Many thanks for sharing the results here! This ablation study is very insightful. I was expecting the mirrored sampling to work better than the vanilla version, at least in terms of the learning speed, but it seems like not. Two questions: 1) Do you also make the `sigma` 3X when you make the `alpha` (learning rate) 3X? 2) Are you using the accelerate `vLLM` version to do the inference or the original `huggingface` implementation?

**Reply by Green0-0** [reply]
I do not make sigma 3X. The vLLM version is used to do inference. Also, its possible my enhancements would have more impact, but the model seems stuck at a cap of 0.6 reward.

**Reply by VsonicV** [reply]
True, but the convergence curve also gives us some insights about the effectiveness of these potential enhancements.
